### ISSUE TITLE
support input methods, pls

### ISSUE THREAD

**Issue by maichaide** [dail8859/NotepadNext#218]
If possible, please consider supporting input methods under linux, such as fcitx and ibus, which can't input Chinese at present.

**Comment by dail8859** [comment]
I'm not familiar with fcitx, ibus, or dealing with any types of input methods. Not sure if this is something that needs adjusted in Qt or Scintilla currently.

**Comment by NichtsHsu** [comment]
我印象中这是 Qt 的问题，如果是 fcitx，需要把 libfcitxplatforminputcontextplugin.so 放在某个目录去加载。具体的细节不太记得了

**Comment by dail8859** [comment]
@NichtsHsu Thanks for the info. Scintilla seems to have some sort of support for IME events, but not sure myself how to test it. Hopefully someone more familiar with it can give it a shot.

**Comment by imkos** [comment]
try https://aur.archlinux.org/packages/fcitx5-qt6-git, I succeeded.

**Comment by dail8859** [comment]
@imkos Thanks for looking into this! There isn't much information about fcitx and Qt, but the only thing I found was exactly what you suggested, and that is to [update fcitx](https://forum.qt.io/post/687735).

So I guess the question is now is if this project is able to do anything to address this? Or best to leave it up to the user (and possibly document the solution somewhere).

**Comment by NichtsHsu** [comment]
Hi everybody, here is a way to build NotepadNext with fcitx support.

1. Clone [fcitx-qt5](https://github.com/fcitx/fcitx-qt5.git).
2. Install `extra-cmake-modules` package.
3. `export` your Qt toolkit to `$PATH`, in my case, it means:

    ```bash
    export PATH=~/Qt/6.4.0/gcc_64/bin:~/Qt/Tools/CMake/bin:$PATH
    ```

4. Enter fcitx-qt5 folder.
5. If you are using Qt6 like me, modify the `CMakeLists.txt` as following:

    ```diff
    diff --git a/CMakeLists.txt b/CMakeLists.txt
    index f9c95af..9358528 100644
    --- a/CMakeLists.txt
    +++ b/CMakeLists.txt
    @@ -10,8 +10,8 @@ find_package(ECM 1.4.0 REQUIRED NO_MODULE)
 
     set(CMAKE_MODULE_PATH ${ECM_MODULE_PATH} ${ECM_KDE_MODULE_DIR} 
    ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
 
    -option(ENABLE_QT5 "Enable Qt5" On)
    -option(ENABLE_QT6 "Enable Qt6 im module" Off)
    +option(ENABLE_QT5 "Enable Qt5" off)
    +option(ENABLE_QT6 "Enable Qt6 im module" on)
     option(ENABLE_LIBRARY "Qt library" On)
 
     include(GNUInstallDirs)
    ````

6. Build it:

    ```bash
    mkdir build && cd build
    cmake ../
    make -j8
    ```

7. Copy the `fcitx-qt5/build/qt6/platforminputcontext/libfcitxplatforminputcontextplugin-qt6.so` to `~/Qt/6.4.0/gcc_64/plugins/platforminputcontexts/` in my case. Should be similar in Qt5.
8. Open Qt Creator, build NotepadNext, now you can use fcitx to input Chinese or others.

**Comment by NichtsHsu** [comment]
I think packaging this library in Appimage allows NotepadNext to support fcitx easily.

**Comment by dail8859** [comment]
@NichtsHsu Thanks! If/when this project moves to cmake then this might be the best solution.

So what is the root cause of the need to package fcitx manually? Do Linux OSs just not have an up-to-date fcitx installed by default?

**Comment by NichtsHsu** [comment]
@dail8859 I guess it might be related with how the AppImage is packaged and executed.

On Ubuntu, `fcitx-qt5` is included in the package `fcitx5-frontend-qt5`. After installing this package, the library file `libfcitxplatforminputcontextplugin.so` will be installed in the `/usr/lib/x86_64-linux-gnu/qt5/plugins/platforminputcontexts/` directory. At this point, use Qt5 to compile NotepadNext in Qt Creator, and you can input Chinese normally when running program. However, when running [your AppImage](https://github.com/dail8859/NotepadNext/releases/download/v0.5.5/NotepadNext-x86_64.AppImage), fcitx is still not responding.

I observed several Qt programs on my PC and found that some of them installed this library file in their own directory:
```bash
$ locate libfcitxplatforminputcontextplugin.so
/opt/kingsoft/wps-office/office6/qt/plugins/platforminputcontexts/libfcitxplatforminputcontextplugin.so
/opt/zoom/platforminputcontexts/libfcitxplatforminputcontextplugin.so
```

I'm not sure why, but they should have a reason for doing so.

**Comment by wineee** [comment]
> I'm not sure why, but they should have a reason for doing so.

wps use vendor libraries,  they can't install libfcitxplatforminputcontextplugin.so to /usr/lib，otherwise it will conflict with `fcitx-qt5`

wps will set QT_PLUGIN_PATH environment variable when startup，to use this library
 `export QT_PLUGIN_PATH=/opt/kingsoft/wps-office/office6/qt/plugins`

**Comment by shenlebantongying** [comment]
The method mentioned by @NichtsHsu is the right direction.

I added fcitx's support to another Qt6 app and it works:

https://github.com/xiaoyifang/goldendict-ng/pull/802

---

`linuxdeployqt` will scan the qt's `platforminputcontexts` folder.

https://github.com/probonopd/linuxdeployqt/blob/8428c59318b250058e6cf93353e2871072bbf7f9/tools/linuxdeployqt/shared.cpp#L1379-L1381

Just install the `libfcitx5platforminputcontextplugin.so` to the corresponding qt's dir, and let `linuxdeployqt` to do the rest.

---

If `AppDir/usr/plugins/platforminputcontexts/libfcitx5platforminputcontextplugin.so` exist in the built appimage, then fcitx should work.

P.S. I googled related problems and I somehow came to this issue. I don't even know what is this project. Hopefully, my PR can be a reference if someone wants to add related code to this project's build script `:)`

**Comment by shenlebantongying** [comment]
> So what is the root cause of the need to package fcitx manually? 

No, not bundle fcitx entirely, just an im (input method) plugin (which I don't know in great detail, the app needs to load it so that fcitx can communicate with the app?)

There is probably already an ibus plugin under `AppDir/usr/plugins/platforminputcontexts/`.

> Do Linux OSs just not have an up-to-date fcitx installed by default?

Yes, they have, but appimage needs the lowest version possible, and ubuntu20.04 don't have fcitx's qt6 plugin, so we have to build one manually (or bump the ubuntu version to 22.10 and just `sudo apt-get install fcitx5-frontend-qt6` in the CI).

**Comment by shenlebantongying** [comment]
Simple, just don't use AppImage which is broken here and there for corner cases like this issue.

Learn to use build tools and build from source.

**Comment by lnxyz** [comment]
我是用源码直接到src下面进行qmake6 NotepadNext.pro &&  make -j$(nproc) 这样把v0.13 编译出来 再进行appimage打包时把系统的so `cp /usr/lib/x86_64-linux-gnu/qt6/plugins/platforminputcontexts/libfcitx* AppDir/usr/plugins/platforminputcontexts/ ` 放到 appimage里面，就可以用fcitx了。NotepadNext默认打包出来是只有ibus 的，所以用fcitx的用户要自己动手了。
